feat(ExpandCollapse): introduce ExpandCollapse component

mareklibra opened the pull request:
<!-- What changes are being made? (What issue is being addressed here?) -->
**What**: introduce ExpandCollapse component

<!-- Please provide a link to your fork's Storybook. See README notes on how to do this. -->
**Link to Storybook**: [https://rawgit.com/mareklibra/patternfly-react/ExpandCollapse-sb4/index.html](https://rawgit.com/mareklibra/patternfly-react/ExpandCollapse-sb4/index.html?knob-bordered=true&knob-textExpanded=Hide%20Advanced%20Options&knob-textCollapsed=Show%20Advanced%20Options&selectedKind=patternfly-react%2FForms%20and%20Controls%2FExpand%20Collapse&selectedStory=ExpandCollapse&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

<!-- Are there any upstream issues or separate issues you need to reference? -->
**Additional issues**:

**Additional info**:
Implementation for the Expand Collapse Section is provided.
[Patternfly.org Design](https://www.patternfly.org/pattern-library/forms-and-controls/expand-collapse-section/)

The component can be farther used for the "More Information" use-case,
where additional text/components are hidden unless the user clicks
on a link.

Such a per-request rendered content is used for texts exceeding scope of a tooltip or
when the user is expected to copy and paste part of the information.

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/src/components/ExpandCollapse/ExpandCollapse.stories.js:
```diff
@@ -0,0 +1,29 @@
+import { ExpandCollapse } from './index';
+import { ALIGN_TYPES } from './constants';
+
+const stories = storiesOf(`${storybookPackageName(name)}/${STORYBOOK_CATEGORY.WIDGETS}/Expand Collapse`, module);
```
This should be under `STORYBOOK_CATEGORY .FORMS_AND_CONTROLS`

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/src/components/ExpandCollapse/ExpandCollapse.stories.js:
```diff
@@ -0,0 +1,29 @@
+);
+
+stories.add(
+  'ExpandCollapse',
```
Would show the left/center better if we added a div around this with a set width and a border: 
```
<div style={{ width: '400px', border: '1px solid lightgray' }}
  <ExpandCollapse  ...
```

mareklibra replied:
done

mareklibra commented:
@serenamarie125 , sure, I will change the underline and add property for text of the expanded state.

Regarding default text values and moving under Forms and Controls:
The component is expected to have multiple use-cases, so far identified are 
- More Information 
- or Expand Collapse Section for forms.

I can imagine more coming in the future.

We can either keep the component under the WIDGETS and let the user just provide proper properties or we can implement specialized components for each of these use cases by providing expected texts or alignment. What do you think? @jeff-phillips-18 , @serenamarie125

mareklibra commented:
Changes: the link consists of text only (not the space or icon). Consuming code can provide props for both expanded and collapsed state.

Storybook is updated (but requires cleaning of browser cache).

jeff-phillips-18 commented:
I think that since the text can be set by the application there is no need for multiple components. I still feel this belongs under Forms and Controls to be consistent with ExpandCollapse pattern found on Patternfly.org. More Information is really just expand/collapse isn't it?

Also, can we set the button's hover state to use `outline: 0`. The focus treatment is really over done. This should be changed generically but in this case it seems really bad to me. The underline is enough to indicate focus in my opinion.

serenamarie125 commented:
@mareklibra the storybook link doesn't work for me now.  Can you update that so I can approve?  As long as we can modify the text & there is the ability to have 2 different labels based on if it's expanded or collapsed we should be fine.

Agree that it likely needs to be in Forms & Controls

mareklibra commented:
The storybook worked for me once browser's cache is cleared. Anyway, redeployed to a new branch, can you please have a look?

Changes:
- moved under Forms & Controls
- so default texts changed to Hide/Show Information
- the button has `outline:none` on focus

Regarding `More Information`: the difference is just in text and alignment of the link.

serenamarie125 commented:
@mareklibra I have a couple of more asks ... 
1 - the caret should be "selectable" as well as the text
2 - right now the text jumps when the component is selected, I think @jeff-phillips-18 may have some thoughts on why
3 - we should have a line to the right of the link, IMO this should be optional, but shown by default

![](https://www.patternfly.org/pattern-library/forms-and-controls/expand-collapse-section/img/advanced-options-collapsed.png)

mareklibra commented:
@jeff-phillips-18 , thanks, good idea with the use of `flex`.

jeff-phillips-18 commented on packages/patternfly-3/patternfly-react/less/expand-collapse.less:
```diff
@@ -0,0 +1,15 @@
+      outline: none;
+    }
+  }
+}
```
This file needs to be updated to what changed in the sass version.

mareklibra replied:
done

mareklibra commented:
Storybook updated

jeff-phillips-18 commented:
@mareklibra I'm still not seeing the updates in the storybook

mareklibra commented:
@jeff-phillips-18 , please excuse late answer. 
The link in the PR description has been updated, storybook should work now.
The PR is rebased.

priley86 commented:
hey @mareklibra - this looks excellent to me... I have one suggestion for allowing the consumer to explicitly override the expand/collapse if desired. I think the "controlled" pattern for simple UI state is a nice addition to a component like this. You can see the behavior in my storybook w/ new knob [here](https://rawgit.com/priley86/patternfly-react/expand-collapse-storybook/index.html?knob-bordered=true&knob-textExpanded=Hide%20Advanced%20Options&knob-textCollapsed=Show%20Advanced%20Options&selectedKind=patternfly-react%2FForms%20and%20Controls%2FExpand%20Collapse&selectedStory=ExpandCollapse&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

Do you mind cherry-picking?

https://github.com/priley86/patternfly-react/commit/5bab582ac052f7a62aaadb50201179c59db18edf

mareklibra commented:
@priley86 , thanks, cherry-picked.

Last changes:
- rebased
- cherry-picked https://github.com/priley86/patternfly-react/commit/5bab582ac052f7a62aaadb50201179c59db18edf
- added `expanded` to `propTypes` and `defaultProps`

priley86 commented:
great! looks good @mareklibra - one last favor... can you update the snapshots? just run `yarn test -u` and check them in to pass the build...

lizsurette commented:
@mareklibra I know I'm late to the party but just wanted to say this looks great! Thanks so much for the contribution and looking forward to making use of it :)

mareklibra commented:
Thanks for the nice feedback :-) I have updated the test snapshot for the `expanded` prop.
